src/ValueParsers/TimeParser.php

```diff
@@ -43,83 +44,113 @@ class TimeParser extends StringValueParser {
 	 * @param ParserOptions|null $options
 	 */
 	public function __construct( CalendarModelParser $calendarModelParser, ParserOptions $options = null ) {
-
-		$options->defaultOption( TimeParser::OPT_CALENDAR, TimeParser::CALENDAR_GREGORIAN );
-		$options->defaultOption( TimeParser::OPT_PRECISION, TimeParser::PRECISION_NONE );
-
 		parent::__construct( $options );
+
 		$this->calendarModelParser = $calendarModelParser;
+
+		$this->defaultOption( self::OPT_CALENDAR, self::CALENDAR_GREGORIAN );
+		$this->defaultOption( self::OPT_PRECISION, self::PRECISION_NONE );
 	}
 
+	/**
+	 * @param string $value
+	 *
+	 * @throws InvalidArgumentException
+	 * @throws ParseException
+	 * @return TimeValue
+	 */
 	protected function stringParse( $value ) {
-		$timeParts = $this->splitTimeString( $value );
-		$timeParts['year'] = $this->padYear( $timeParts['year'] );
-
-		$calendarOpt = $this->getOptions()->getOption( TimeParser::OPT_CALENDAR );
-		$calendarModelRegex = '/(' . preg_quote( self::CALENDAR_GREGORIAN, '/' ). '|' . preg_quote( self::CALENDAR_JULIAN, '/' ) . ')/i';
-
-		if( $timeParts['calendar'] === '' && preg_match( $calendarModelRegex, $calendarOpt ) ) {
-			$timeParts['calendar'] = $calendarOpt;
-		} else if( $timeParts['calendar'] !== '' ) {
-			$timeParts['calendar'] = $this->calendarModelParser->parse( $timeParts['calendar'] );
-		} else {
-			$timeParts['calendar'] = self::CALENDAR_GREGORIAN;
+		if ( !is_string( $value ) ) {
+			throw new InvalidArgumentException( '$value must be a string' );
 		}
 
-		$precisionOpt = $this->getOptions()->getOption( TimeParser::OPT_PRECISION );
-		$precisionFromTime = $this->getPrecisionFromTimeParts( $timeParts );
-		if( is_int( $precisionOpt ) && $precisionOpt <= $precisionFromTime ) {
-			$precision = $precisionOpt;
-		} else {
-			$precision = $precisionFromTime;
-		}
+		$timeParts = $this->splitTimeString( $value );
+		// Pad sign with 1 plus, year with 16 zeros and hour, minute and second with 2 zeros
+		$time = vsprintf( '%\'+1s%016s-%s-%sT%02s:%02s:%02sZ', $timeParts );
+		$precision = $this->getPrecision( $timeParts );
+		$calendarModel = $this->getCalendarModel( $timeParts[7] );
 
-		$time = $this->getTimeStringFromParts( $timeParts );
 		try {
-			return new TimeValue( $time, 0, 0, 0, $precision, $timeParts['calendar'] );
+			return new TimeValue( $time, 0, 0, 0, $precision, $calendarModel );
```

Comment:
Shouldn't start to default "after" to 1? Or was it "before"? I'm still confused about which is which... but one of them should be 1, not 0.